File: app/javascript/channels/room_channel.js
import consumer from "./consumer"

// $(function() { ... }); で囲むことでレンダリング後に実行される
// レンダリング前に実行されると $('#messages').data('room_id') が取得できない
// turbolinks を使っている場合は $(document).on('turbolinks:load', function() { ... }); で囲うorturbolinksの設定を無効にする。
// $(document).on('turbolinks:load', function() {
  $(function(){
      const chatChannel = consumer.subscriptions.create({ channel: 'RoomChannel', room: $('#messages').data('room_id') }, {
        connected() {
          // Called when the subscription is ready for use on the server
        },
  
        disconnected() {
          // Called when the subscription has been terminated by the server
        },
        
        received: function(data) {
          return $('#messages').append(data['message']);
        },
  
        speak: function(message) {
          return this.perform('speak', {message: message});
        }
  
      });
  
      $(document).on('keypress', '[data-behavior~=room_speaker]', function(event) {
        if (event.key === 'Enter') {
          chatChannel.speak(event.target.value);
          event.target.value = '';
          return event.preventDefault();
        }
      });
  });
// });

File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  
  def subscribed
    stream_from "room_channel_#{params['room']}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    if current_user.room_users.present? && current_user.room_users.find_by(room_id: params['room'])
      Message.create! content: data['message'], user_id: current_user.id, room_id: params['room']
    end
  end
end
